src/Drivers/miniqmc_sync_move.cpp

```diff
@@ -443,6 +445,7 @@ int main(int argc, char** argv)
       const std::vector<Mover*> Sub_list(extract_sub_list(mover_list, first, last));
       const std::vector<ParticleSet*> P_list(extract_els_list(Sub_list));
       const std::vector<WaveFunction*> WF_list(extract_wf_list(Sub_list));
+      const std::vector<NonLocalPP<RealType>*> NLPP_list(extract_nlpp_list(Sub_list));
```

Comment:
Doing this every batch seems pointless.

Reply:
There will be an optimal batch size for the entire block for a particular system there is no need to vary it or recalculate it here.

Reply:
If you implement dynamic population DMC, the batch size can vary. Some batches may have fewer walkers.

Reply:
That's not a good idea.  Some batches could have fewer walkers but the size of the batch scope should remain invariant. You could not dispatch computation over the empty/dead walkers data but there is nothing but disadvantage to having that allocation be dynamic at the QMC block level. Walkers transfer should never happen except for new walker spawning. The "load" imbalance between CPU threads feeding a particular GPU is not important since these are async calls. If it did matter for CPU only that should be a "specialization" for CPU.

Reply:
Obviously you need an algorithm for spawing walkers that first puts them into crowds with the most empty slots.

Reply:
Ideally you want to have these DMC crowds out of lockstep as well so a modification to the algorithm could be a big help. If you could use the average energy of the last step modified by the just the current crowds energy or something similar.

Reply:
We do need a resize capability because if the population goes up, buffers need to resize larger. Resize cost is only paid when the size goes larger. Using a separate size and capacity needs a  lot of book keeping and can be dangerous. See our CUDA code, it always need to keep an eye on some sort of mapping tables. If we need to do load balancing between nodes, we need to do the same balancing between crowds. We do kill and spawn walkers. `The "load" imbalance between CPU threads feeding a particular GPU is not important` I can consider a specialization of GPU particularly in the one GPU per MPI scenario. Imbalance problem usually harms all cases.

Reply:
Don't care about the old CUDA code, that's not a viable way to write a good implementation now. I think many things that are too "hard" are because of how the code is written without good encapsulation, separation of concerns, or leveraging the power of the modern c++ type system.  

Copying walkers needlessly between allocations will be much worse than all but the worst computational imbalances.  
If you dispute this then we need a miniqmc where both algorithms can both be implemented side by side, i.e. same executable as much as possible the same infrastructure.

Load balance between nodes just kill from the fullest crowds and add to the emptiest crowds doesn't seem difficult.

Reply:
I think this larger DMC implementation conversation is useful but too big for here.

Reply:
Even in the case of fixed population. We need to kill walkers with small weights and spawn walkers from walkers with large weights within a node and between nodes. Copy walkers is not avoidable and help reducing the imbalance. It is faster than recomputing. Copying or recomputing walkers anyway add imbalance and thus the faster the better.